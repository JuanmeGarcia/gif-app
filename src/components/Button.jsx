import React from 'react'

function Button({handleAdd}) {
	return (
		<button 
            className="button" 
            onClick={handleAdd}
            type="submit"
        >
			Agregar
		</button>
	)
}

export { Button }
