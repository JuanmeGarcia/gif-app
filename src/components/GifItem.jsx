import React from 'react'
import PropTypes from 'prop-types';

function GifItem({ url, title }) {
	return (
        <div className='card'>
            <img className='gif' data-testid='gif-image' src={url} alt={title} />
            <p data-testid='gif-title'>{title}</p>
        </div>
    )
}

export { GifItem }


GifItem.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
}