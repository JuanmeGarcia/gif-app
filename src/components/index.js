export * from './AddCategory'
export * from './Button'
export * from './GifGrid'
export * from './GifItem'