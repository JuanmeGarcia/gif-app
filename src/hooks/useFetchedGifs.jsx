import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";


export function useFetchedGifs(category) {
    const [gifs, setGifs] = useState([])
    const [loading, setLoading] = useState(true)

    useEffect(()=>{
        setTimeout(()=>{
            getGifs(category)
                .then(gifs => setGifs(gifs))
            
            setLoading(false)
        }, 1000)
    }, [])

    return {
        gifs,
        loading
    }
}