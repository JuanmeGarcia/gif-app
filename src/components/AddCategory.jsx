import React, { useRef, useState } from 'react'
import {Button} from './Button'
import PropTypes from 'prop-types'

function AddCategory({ onAddCategory }) {
	const [text, setText] = useState('')
    const inputRef = useRef(null)

	const onChange = e => {
		setText(e.target.value)
	}

	const handleSubmit = e => {
        e.preventDefault()
        const inputValue = text.trim()
        if(inputValue.length > 2) {
            onAddCategory(inputValue)
            setText('')
            inputRef.current.focus()
        }else{
            alert('El nombre de la categoria debe tener al menos 3 caracteres')
        }

	}

	return (
		<form 
            action="" 
            onSubmit={handleSubmit}
            aria-label="form"
        >
			<input 
                type="text" 
                value={text} 
                onChange={onChange} 
                ref={inputRef}
            />
            <Button />
		</form>
	)
}
AddCategory.propTypes = {
    onAddCategory: PropTypes.func.isRequired
}


export { AddCategory }

