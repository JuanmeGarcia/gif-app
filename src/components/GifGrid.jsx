import { useFetchedGifs } from '../hooks/useFetchedGifs'
import { GifItem } from './GifItem'
import { PropTypes } from 'prop-types'

function GifGrid({ category }) {
	const { gifs, loading } = useFetchedGifs(category)

	const listImages = gifs.map((img, i) => {
		return (
			<GifItem 
				key={i}
				title={img.title}
				url={img.url}
			/> //mala practica poner key={i}
		)
	})

	return (
		<>
			<h3>{category}</h3>
			{loading && <h2 className="loading">Loading...</h2>}
			<section className="gif-grid">{listImages}</section>
		</>
	)
}

GifGrid.propTypes = {
	category: PropTypes.string.isRequired
}

export { GifGrid }
