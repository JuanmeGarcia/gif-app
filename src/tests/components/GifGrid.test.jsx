import { render, screen } from '@testing-library/react'
import { GifGrid } from '../../components'
import { useFetchedGifs } from '../../hooks/useFetchedGifs'

jest.mock('../../hooks/useFetchedGifs')

describe('Test suite of <GifGrid />', () => {

    const category = 'One Punch'
    const dummyGifs = [
        {
            id: 'abc',
            title: 'saitama',
            url: 'https://localhost/saitama.jpg'
        },
        {
            id: 'def',
            title: 'goku',
            url: 'https://localhost/goku.jpg'
        }
    ] 

    test('should show a loading initially', () => {

        useFetchedGifs.mockReturnValue({
            gifs: [],
            loading: true
        })

        render(
            <GifGrid 
                category={category}
            ></GifGrid>
        )
        screen.debug()

        const text = screen.getByText('Loading...')
        const title = screen.getByText(category)
        
        expect(text).toBeTruthy()
        expect(title.innerHTML).toBe(category)
    })

    test('should show items when images are load by custom hook', () => {

        useFetchedGifs.mockReturnValue({
            gifs: dummyGifs,
            loading: false
        })

        const { container } = render(
            <GifGrid 
                category={category}
            ></GifGrid>
        )
        screen.debug()

        const images = screen.getAllByRole('img')
        expect(container).toMatchSnapshot()
        expect(images.length).toBe(2)
    })
})