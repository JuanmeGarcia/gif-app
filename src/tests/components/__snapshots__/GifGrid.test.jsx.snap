// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite of <GifGrid /> should show items when images are load by custom hook 1`] = `
<div>
  <h3>
    One Punch
  </h3>
  <section
    class="gif-grid"
  >
    <div
      class="card"
    >
      <img
        alt="saitama"
        class="gif"
        data-testid="gif-image"
        src="https://localhost/saitama.jpg"
      />
      <p
        data-testid="gif-title"
      >
        saitama
      </p>
    </div>
    <div
      class="card"
    >
      <img
        alt="goku"
        class="gif"
        data-testid="gif-image"
        src="https://localhost/goku.jpg"
      />
      <p
        data-testid="gif-title"
      >
        goku
      </p>
    </div>
  </section>
</div>
`;
